File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/stories/StoryAddSheet.kt
/*
 * StoryAddSheet.kt
 *
 * The "+ Add" bottom sheet from §7.3a of the v3 mockup. Three
 * sections — capture / library / layout — each with icon + label +
 * hint rows. Mirror of iOS `StoryAddSheet.swift`.
 */

@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package app.quickink.mobile.features.stories

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.Notes
import androidx.compose.material.icons.outlined.CalendarToday
import androidx.compose.material.icons.outlined.CameraAlt
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Edit
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material.icons.outlined.Mic
import androidx.compose.material.icons.outlined.Photo
import androidx.compose.material.icons.filled.LocationOn
import androidx.compose.material.icons.outlined.TextFields
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.data.storyitem.StoryItemEntity
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkSpacing
import kotlinx.coroutines.delay

@Composable
fun StoryAddSheet(
    precedingItemCaption: String?,
    /** User ID — threaded down so the library picker can scope its
     *  capture query. */
    userId: String,
    onPickInlineKind: (StoryItemEntity.Kind) -> Unit,
    onPickVoiceClip: (audioUri: String, durationMs: Long) -> Unit,
    /** Phase 2 follow-up — capture-backed item insertion. */
    onPickCapture: (captureId: String, kind: StoryItemEntity.Kind) -> Unit,
    onDismiss: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    val sheetState = rememberModalBottomSheetState()

    var showVoiceRecorder by remember { mutableStateOf(false) }
    var pickerFilter by remember { mutableStateOf<StoryPickerFilter?>(null) }
    var showNotePicker by remember { mutableStateOf(false) }
    var stubToast by remember { mutableStateOf<String?>(null) }

    fun showStub(message: String) {
        stubToast = message
    }

    LaunchedEffect(stubToast) {
        if (stubToast != null) {
            delay(1_800)
            stubToast = null
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState       = sheetState,
        containerColor   = colors.surface,
    ) {
        Column(
            modifier = Modifier.padding(
                start  = QuickInkSpacing.s4,
                end    = QuickInkSpacing.s4,
                bottom = QuickInkSpacing.s6,
            ),
        ) {
            Text("Add", style = type.editorial, color = colors.ink)
            Text(
                text = buildAnnotatedString {
                    if (precedingItemCaption.isNullOrEmpty()) {
                        append("at the start of the story")
                    } else {
                        append("after — ")
                        withStyle(SpanStyle(fontStyle = FontStyle.Italic)) {
                            append("\"")
                            append(precedingItemCaption)
                            append("\"")
                        }
                    }
                },
                style = type.bodyItalic,
                color = colors.inkSoft,
                modifier = Modifier.padding(bottom = QuickInkSpacing.s3),
            )

            SectionHeader("CAPTURE NEW")
            AddRow(icon = Icons.Outlined.CameraAlt,  label = "Scan a page",         hint = "camera",     onClick = { showStub("Scan a page coming soon") })
            AddRow(icon = Icons.Outlined.Image,      label = "Take a photo",        hint = "camera",     onClick = { showStub("Take a photo coming soon") })
            AddRow(icon = Icons.Outlined.Mic,        label = "Record a voice note", hint = "tap & hold", onClick = { showVoiceRecorder = true })

            SectionDivider()
            SectionHeader("FROM YOUR LIBRARY")
            AddRow(icon = Icons.Outlined.Photo,      label = "Choose a photo",      hint = "picker",     onClick = { pickerFilter = StoryPickerFilter.PHOTO })
            AddRow(icon = Icons.Outlined.Description, label = "Choose a document",  hint = "picker",     onClick = { pickerFilter = StoryPickerFilter.DOCUMENT })
            AddRow(icon = Icons.AutoMirrored.Outlined.Notes, label = "Choose a note", hint = "picker",   onClick = { showNotePicker = true })

            SectionDivider()
            SectionHeader("LAYOUT")
            AddRow(icon = Icons.Outlined.TextFields,  label = "Write a paragraph",  hint = "serif",      onClick = { onPickInlineKind(StoryItemEntity.Kind.TEXT_BLOCK) })
            AddRow(icon = Icons.Outlined.Edit,        label = "Handwritten note",   hint = "Caveat",     onClick = { onPickInlineKind(StoryItemEntity.Kind.HANDWRITTEN_NOTE) })
            AddRow(icon = Icons.Outlined.CalendarToday, label = "Date divider",     hint = "— May 5 —",  onClick = { onPickInlineKind(StoryItemEntity.Kind.DATE_DIVIDER) })
            AddRow(icon = Icons.Filled.LocationOn,    label = "Place pin",          hint = "Shibuya",    onClick = { onPickInlineKind(StoryItemEntity.Kind.PLACE_PIN) })

            if (stubToast != null) {
                Text(
                    text     = stubToast!!,
                    style    = type.bodyItalic,
                    color    = colors.inkSoft,
                    modifier = Modifier.padding(top = QuickInkSpacing.s2),
                )
            }
        }
    }

    if (showVoiceRecorder) {
        StoryVoiceClipRecorderSheet(
            onSave   = { uri, durationMs ->
                showVoiceRecorder = false
                onPickVoiceClip(uri, durationMs)
            },
            onCancel = { showVoiceRecorder = false },
        )
    }

    val currentPicker = pickerFilter
    if (currentPicker != null) {
        StoryLibraryPickerSheet(
            userId    = userId,
            filter    = currentPicker,
            onPick    = { captureId ->
                val kind = if (currentPicker == StoryPickerFilter.PHOTO)
                    StoryItemEntity.Kind.PHOTO else StoryItemEntity.Kind.DOCUMENT
                pickerFilter = null
                onPickCapture(captureId, kind)
            },
            onDismiss = { pickerFilter = null },
        )
    }

    if (showNotePicker) {
        StoryNotePickerSheet(
            userId    = userId,
            onPick    = { entryId ->
                showNotePicker = false
                onPickCapture(entryId, StoryItemEntity.Kind.NOTE)
            },
            onDismiss = { showNotePicker = false },
        )
    }
}

@Composable
private fun SectionHeader(label: String) {
    val colors = LocalQuickInkColors.current
    Text(
        text          = label,
        color         = colors.muted,
        fontSize      = 11.sp,
        fontWeight    = FontWeight.Medium,
        letterSpacing = 1.5.sp,
        modifier      = Modifier.padding(top = QuickInkSpacing.s2, bottom = QuickInkSpacing.s1),
    )
}

@Composable
private fun SectionDivider() {
    val colors = LocalQuickInkColors.current
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(0.5.dp)
            .background(colors.borderSoft)
            .padding(vertical = QuickInkSpacing.s1),
    )
}

@Composable
private fun AddRow(
    icon: ImageVector,
    label: String,
    hint: String,
    onClick: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(vertical = QuickInkSpacing.s2),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(QuickInkSpacing.s3),
    ) {
        Box(
            modifier = Modifier
                .size(36.dp)
                .clip(RoundedCornerShape(8.dp))
                .background(colors.borderSoft),
            contentAlignment = Alignment.Center,
        ) {
            Icon(imageVector = icon, contentDescription = null, tint = colors.ink, modifier = Modifier.size(16.dp))
        }
        Text(text = label, style = type.body, color = colors.ink)
        Spacer(modifier = Modifier.weight(1f))
        Text(text = hint, style = type.bodyItalic, color = colors.muted, fontSize = 12.sp)
    }
}

File: apps/quickink/ios/QuickInk/Stories/StoryAddSheet.swift
/*
 * StoryAddSheet.swift
 *
 * The "+ Add" bottom sheet from §7.3a of the v3 mockup. Three
 * sections — capture / library / layout — each with icon + label +
 * hint rows. The italic subtitle anchors the insertion point with
 * "after — \"<preceding caption>\"".
 *
 * Phase 2 scope:
 *   - Inline kinds (text_block, handwritten_note, date_divider,
 *     place_pin): fully functional — `onPickInlineKind`.
 *   - Voice clip: opens an in-sheet tap-and-hold recorder
 *     (`StoryVoiceClipRecorderView`); on release commits via
 *     `onPickVoiceClip`. Local-only AAC-LC 64 kbps mono, max 10 s.
 *     Drive sync wires up via the dirty flag set on insert.
 *   - Capture (scan / take photo) + library pickers stub to
 *     `onPickStubbed` (a no-op closer) — Phase 2 follow-up.
 *
 * Mirror of Android `StoryAddSheet.kt`.
 */

import SwiftUI

struct StoryAddSheet: View {

    let precedingItemCaption: String?
    /// User ID — threaded down so the library picker can scope its
    /// capture query.
    let userId: String
    var onPickInlineKind: (StoryItem.Kind) -> Void
    var onPickVoiceClip: (_ audioUri: String, _ durationMs: Int) -> Void
    /// Phase 2 follow-up — fires after the user picks a capture from
    /// the library picker. `kind` discriminates "photo"
    /// (`StoryItem.Kind.photo`) vs "document" (`.document`); both
    /// flow into the same story_item row but carry their semantic
    /// kind for renderer dispatch.
    var onPickCapture: (_ captureId: String, _ kind: StoryItem.Kind) -> Void
    var onPickStubbed: () -> Void = {}

    @State private var showVoiceRecorder = false
    @State private var pickerFilter: StoryLibraryPickerSheet.Filter? = nil
    @State private var showNotePicker = false
    @State private var stubToast: String? = nil

    var body: some View {
        VStack(spacing: 0) {
            handle
            header
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: QuickInkSpacing.s2) {
                    sectionHeader("CAPTURE NEW")
                    row(icon: "camera",                 label: "Scan a page",          hint: "camera",       action: stub("Scan a page coming soon"))
                    row(icon: "photo",                  label: "Take a photo",         hint: "camera",       action: stub("Take a photo coming soon"))
                    row(icon: "mic",                    label: "Record a voice note",  hint: "tap & hold",   action: { showVoiceRecorder = true })

                    sectionDivider()
                    sectionHeader("FROM YOUR LIBRARY")
                    row(icon: "photo.on.rectangle",     label: "Choose a photo",       hint: "picker",       action: { pickerFilter = .photo })
                    row(icon: "doc.text",               label: "Choose a document",    hint: "picker",       action: { pickerFilter = .document })
                    row(icon: "note.text",              label: "Choose a note",        hint: "picker",       action: { showNotePicker = true })

                    sectionDivider()
                    sectionHeader("LAYOUT")
                    row(icon: "textformat",             label: "Write a paragraph",    hint: "serif",        action: { onPickInlineKind(.textBlock) })
                    row(icon: "pencil",                 label: "Handwritten note",     hint: "Caveat",       action: { onPickInlineKind(.handwrittenNote) })
                    row(icon: "calendar",               label: "Date divider",         hint: "— May 5 —",    action: { onPickInlineKind(.dateDivider) })
                    row(icon: "mappin.and.ellipse",     label: "Place pin",            hint: "Shibuya",      action: { onPickInlineKind(.placePin) })
                }
                .padding(.horizontal, QuickInkSpacing.s4)
                .padding(.bottom, QuickInkSpacing.s6)
            }
            if let toast = stubToast {
                Text(toast)
                    .font(QuickInkText.bodyItalic)
                    .foregroundStyle(QuickInkColors.inkSoft)
                    .padding(.horizontal, QuickInkSpacing.s4)
                    .padding(.vertical, QuickInkSpacing.s2)
            }
        }
        .background(QuickInkColors.surface)
        .sheet(isPresented: $showVoiceRecorder) {
            StoryVoiceClipRecorderView(
                onSave: { uri, durationMs in
                    showVoiceRecorder = false
                    onPickVoiceClip(uri, durationMs)
                },
                onCancel: { showVoiceRecorder = false }
            )
            .presentationDetents([.medium])
        }
        .sheet(item: Binding(
            get: { pickerFilter.map { PickerFilterIdentifier(filter: $0) } },
            set: { newValue in pickerFilter = newValue?.filter }
        )) { identifier in
            StoryLibraryPickerSheet(
                userId: userId,
                filter: identifier.filter,
                onPick:   { captureId in
                    let kind: StoryItem.Kind = identifier.filter == .photo ? .photo : .document
                    pickerFilter = nil
                    onPickCapture(captureId, kind)
                },
                onCancel: { pickerFilter = nil }
            )
            .presentationDetents([.medium, .large])
        }
        .sheet(isPresented: $showNotePicker) {
            StoryNotePickerSheet(
                userId: userId,
                onPick:   { entryId in
                    showNotePicker = false
                    onPickCapture(entryId, .note)
                },
                onCancel: { showNotePicker = false }
            )
            .presentationDetents([.medium, .large])
        }
    }

    /// SwiftUI's `.sheet(item:)` needs an Identifiable wrapper; the
    /// raw enum can't be `Identifiable` cleanly (no stable `id` for
    /// the case payload). This thin struct gives us one.
    private struct PickerFilterIdentifier: Identifiable {
        let filter: StoryLibraryPickerSheet.Filter
        var id: String { filter.rawValue }
    }

    // MARK: - Pieces

    private var handle: some View {
        Capsule()
            .fill(QuickInkColors.border)
            .frame(width: 38, height: 4)
            .padding(.top, QuickInkSpacing.s2)
            .padding(.bottom, QuickInkSpacing.s3)
    }

    private var header: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text("Add")
                .font(QuickInkText.editorial)
                .foregroundStyle(QuickInkColors.ink)
            if let caption = precedingItemCaption, !caption.isEmpty {
                (Text("after — ") +
                 Text("\"\(caption)\"").italic())
                    .font(QuickInkText.bodyItalic)
                    .foregroundStyle(QuickInkColors.inkSoft)
            } else {
                Text("at the start of the story")
                    .font(QuickInkText.bodyItalic)
                    .foregroundStyle(QuickInkColors.inkSoft)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, QuickInkSpacing.s4)
        .padding(.bottom, QuickInkSpacing.s3)
    }

    private func sectionHeader(_ text: String) -> some View {
        Text(text)
            .font(.system(size: 11, weight: .medium))
            .tracking(1.5)
            .foregroundStyle(QuickInkColors.muted)
            .padding(.top, QuickInkSpacing.s2)
    }

    private func sectionDivider() -> some View {
        Rectangle()
            .fill(QuickInkColors.borderSoft)
            .frame(height: 0.5)
            .padding(.vertical, QuickInkSpacing.s1)
    }

    private func row(icon: String, label: String, hint: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: QuickInkSpacing.s3) {
                ZStack {
                    RoundedRectangle(cornerRadius: 8).fill(QuickInkColors.borderSoft)
                        .frame(width: 36, height: 36)
                    Image(systemName: icon)
                        .font(.system(size: 16))
                        .foregroundStyle(QuickInkColors.ink)
                }
                Text(label)
                    .font(QuickInkText.body)
                    .foregroundStyle(QuickInkColors.ink)
                Spacer()
                Text(hint)
                    .font(QuickInkText.bodyItalic)
                    .foregroundStyle(QuickInkColors.muted)
            }
            .padding(.vertical, QuickInkSpacing.s2)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private func stub(_ message: String) -> () -> Void {
        return {
            stubToast = message
            Task { @MainActor in
                try? await Task.sleep(nanoseconds: 1_800_000_000)
                stubToast = nil
            }
        }
    }
}
